### PR TITLE
Fast abi forwarder tear-off support 

### PR DESCRIPTION
**Pull request by Scottj1s**
Initial support for creating fast abi forwarder:
msbuild props/targets to preprocess and assemble arch-specific thunks
fast_abi_forwarder helper to call into thunks
test collateral

### CODE REVIEW

File: src/tool/cppwinrt/strings/base_fast_fwd.h [hunk 1]
```diff
@@ -0,0 +1,2123 @@
+
+namespace winrt::impl
+{
+// Thunk definitions in arch-specific assembly sources
+#define FAST_FWD_THUNK(index) extern "C" void __stdcall fast_fwd_thunk##index();
```
Comment (kennykerr) [inline src/tool/cppwinrt/strings/base_fast_fwd.h]: Can we call it `WINRT_XXX` for consistency with all the other cppwinrt macros?

File: src/tool/cppwinrt/test_fast_fwd/FastForwarderTests.cpp [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+using namespace Windows::Foundation;
+using namespace FastForwarderTest;
+
+#define REQUIRE_STR_EQUAL(s1, s2) \
+    REQUIRE(std::wstring(s1).compare(std::wstring(s2)) == 0)
+
+template<typename TForwarder, typename TOwner>
+auto make_fast_abi_forwarder(TOwner const& owner, size_t offset)
+{
+	TForwarder forwarder;
```
Comment (kennykerr) [inline src/tool/cppwinrt/test_fast_fwd/FastForwarderTests.cpp]: Some tabs in this file...

File: src/tool/cppwinrt/test_fast_fwd/pch.h [hunk 3]
```diff
@@ -0,0 +1,13 @@
+#pragma once
+
+#include "catch.hpp"
+#include "winrt/coroutine.h"
+#include "winrt/Windows.Foundation.Collections.h"
+#include <windows.h>
+//#include <winrt.h>
```
Comment (kennykerr) [inline src/tool/cppwinrt/test_fast_fwd/pch.h]: Remove

File: src/tool/cppwinrt/test_fast_fwd/pch.h [hunk 4]
```diff
@@ -0,0 +1,13 @@
+#pragma once
+
+#include "catch.hpp"
+#include "winrt/coroutine.h"
```
Comment (kennykerr) [inline src/tool/cppwinrt/test_fast_fwd/pch.h]: Remove

File: src/tool/cppwinrt/test_fast_fwd/pch.h (hunk 3 above)
Comment (kennykerr) [reply src/tool/cppwinrt/test_fast_fwd/pch.h]: Probably a bunch of includes here aren't needed.

File: src/tool/cppwinrt/test_fast_fwd/FastForwarderTest.idl [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+            Double  arg13,
+            Single  arg14,
+            Double  arg15,
+            Single  arg16,
+            Double  arg17,
+            Single  arg18);
+    }
+
+    [contract(FastForwarderTest.InternalContract, 1)]
+    runtimeclass Class
```
Comment (kennykerr) [inline src/tool/cppwinrt/test_fast_fwd/FastForwarderTest.idl]: This class must actually implement multiple interfaces to ensure the forward offsetting works.

Comment (kennykerr) [reply src/tool/cppwinrt/test_fast_fwd/FastForwarderTest.idl]: And the class lacks the [fastabi] attribute so cppwinrt won't know to fold the various interfaces.

File: src/tool/cppwinrt/strings/base_fast_fwd.h (hunk 1 above)
Comment (Scottj1s) [reply src/tool/cppwinrt/strings/base_fast_fwd.h]: yes, WINRT_FF_THUNK

File: src/tool/cppwinrt/test_fast_fwd/FastForwarderTest.idl (hunk 5 above)
Comment (Scottj1s) [reply src/tool/cppwinrt/test_fast_fwd/FastForwarderTest.idl]: right, this was testing the forwarding logic without true multiple interfaces.  will add that in when we add the compiler hook.

**Pull request merged**